core/ledger/ledgermgmt/snapshot_service.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ledgermgmt
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/empty"
+	pb "github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/pkg/errors"
+)
+
+// Snapshot Service implements SnapshotServer grpc interface
+type SnapshotService struct {
+	LedgerGetter LedgerGetter
+}
+
+// LedgerGetter gets the PeerLedger associated with a channel.
+type LedgerGetter interface {
+	GetLedger(cid string) ledger.PeerLedger
+}
+
+// Generate generates a snapshot request.
+func (s *SnapshotService) Generate(ctx context.Context, signedRequest *pb.SignedSnapshotRequest) (*empty.Empty, error) {
```

Comment:
Pardon my interest, but, you have here 
Generate(SignedSnapshotRequest) returns (google.protobuf.Empty) {}
Cancel(SignedSnapshotRequest) returns (google.protobuf.Empty) {}
QueryPendings(SignedSnapshotRequest) returns (QueryPendingSnapshotsResponse) {}

Why don't you make some dispatcher to handle all of them? like in the [discovery service](https://github.com/hyperledger/fabric/blob/23466eb45de730aabdd5491d8ba5e170091d4a52/discovery/service.go#L106-L119) ?

This way you can then add authentication checks and sanity checks only in a single place, and not spread them across this file, and after all checks pass, it will jump to a "function pointer" and return either empty or the response

Reply:
The handler approach causes unnecessary indirection. Which forces you to use `oneof` in  proto request message. IMO, the handler based approach would have been useful if we had a single handler at the peer level that performs these common tasks and new services require just to add new function. While writing full service, I see little value and instead invoke a common function at the beginning of each function sounds to me a good way to avoid indirection.

Reply:
That's not what I am saying. I am not saying to one-of into a single gRPC method, but to just use internally a dispatching pattern, and attach the function pointer in the gRPC function.

Reply:
@yacovm thank you for your comment. We (ledger team) discussed and decided to use multiple rpc methods to avoid using oneof and make each call more expressive.